lib/Elevate/Components/PackageRestore.pm

```diff
@@ -0,0 +1,106 @@
+package Elevate::Components::PackageRestore;
+
+=encoding utf-8
+
+=head1 NAME
+
+Elevate::Components::PackageRestore
+
+Handle restoring packages that get removed during elevate
+
+Before leapp:
+    Detect which packages in our list are installed and
+    store our findings.
+
+After leapp:
+    Reinstall any packages detected pre-leapp
+
+=cut
+
+use cPstrict;
+
+use Elevate::StageFile ();
+
+use Cpanel::Pkgr ();
+
+use File::Copy    ();
+use Log::Log4perl qw(:easy);
+
+use parent qw{Elevate::Components::Base};
+
+#
+# Set as a function for unit testing
+#
+sub _get_packages_to_check () {
+    return qw{
+      net-snmp
+    };
+}
+
+#
+# Set as a function for unit testing
+#
+sub _get_files_to_restore ($package) {
+
+    my $package_files_to_restore = {
+        'net-snmp' => [
+            qw{
+              /etc/snmp/snmpd.conf
+              /etc/snmp/snmptrapd.conf
+            }
+        ],
+    };
+
+    if ( exists $package_files_to_restore->{$package} ) {
+        return $package_files_to_restore->{$package};
+    }
+    else {
+        return [];
+    }
+}
+
+sub pre_leapp ($self) {
+
+    my @package_list = _get_packages_to_check();
+    my @installed_packages;
+
+    foreach my $package (@package_list) {
+        if ( Cpanel::Pkgr::is_installed($package) ) {
+            push @installed_packages, $package;
+        }
+    }
+
+    Elevate::StageFile::update_stage_file(
+        {
+            'packages_to_restore' => \@installed_packages,
+        }
+    );
+
+    return;
+}
+
+sub post_leapp ($self) {
+
+    my $packages = Elevate::StageFile::read_stage_file('packages_to_restore');
+    return unless defined $packages and ref $packages eq 'ARRAY';
+
+    foreach my $package (@$packages) {
+
+        $self->yum->install($package);
+
+        my $restore_files = _get_files_to_restore($package);
+
+        foreach my $restore_file (@$restore_files) {
+            my $backup_file = $restore_file . '.rpmsave';
+
+            if ( -e $backup_file ) {
+                File::Copy::copy( $backup_file, $restore_file )
```

Comment:
Does this preserve the permissions? By default, those files are 0600, but at least in a quick test, I'm getting 0644 files with the copy.

Reply:
I can add a permission check/set to this.

Reply:
Nice catch @sloanebernstein 

https://perldoc.perl.org/File::Copy
```
You may use the syntax use File::Copy "cp" to get at the cp alias for this function. The syntax is exactly the same. The behavior is nearly the same as well: as of version 2.15, cp will preserve the source file's permission bits like the shell utility cp(1) would do with default options, while copy uses the default permissions for the target file (which may depend on the process' umask, file ownership, inherited ACLs, etc.). That is, if the destination file already exists, cp will leave its permissions unchanged; otherwise the permissions are taken from the source file and modified by the umask. If an error occurs in setting permissions, cp will return 0, regardless of whether the file was successfully copied.
```

We should be using `File::Copy::cp()` to perform this task.

@timmullin - Please open a follow up case to audit the rest of the code base for this issue too as I suspect we have this issue elsewhere as well.

Reply:
Created:  https://webpros.atlassian.net/browse/RE-723